llama.h

```diff
@@ -987,7 +987,13 @@ extern "C" {
             struct llama_context * ctx,
           llama_token_data_array * candidates);
 
-    /// @details Randomly selects a token from the candidates based on their probabilities.
+    /// @details Randomly selects a token from the candidates based on their probabilities using a given pointer to a std::mt19937.
+    LLAMA_API llama_token llama_sample_token_with_rng(
+            struct llama_context * ctx,
+          llama_token_data_array * candidates,
+                            void * rng);
```

Comment:
Not ok to pass a pointer to a C++ object. This is supposed to be a C API.

Reply:
Good point. What do you think would be a good alternative way to fix this?

Reply:
Do you think it would make sense to add an API for saving/loading RNG state via `char *`?

Reply:
I can think of two alternatives:
- Keep one rng per sequence in `llama_context`, add a new `llama_sample_token` function that receives a sequence id
- Add a new `llama_sample_token` function that receives to receives a value between 0 and 1 and samples based on that value

Reply:
> Do you think it would make sense to add an API for saving/loading RNG state via `char *`?

I wouldn't like it very much as a solution, it would complicate the API and it would be inefficient, even though in practice it is not likely to have a measurable effect on performance, the RNG state size can be significant.

Reply:
> * Keep one rng per sequence in `llama_context`, add a new `llama_sample_token` function that receives a sequence id. This would also require also updating the state saving code.

In fact I think that only this option will work correctly with the state saving code, so I would prefer this solution.

Reply:
Seems like the RNG state is suitable to be part of the `llama_sampling_context`. Should we fix this issue within the scope of https://github.com/ggerganov/llama.cpp/issues/5214 ?

Reply:
The problem is that the internal state of the RNG is not a simple unsigned int like the seed but rather 19337 bits and the way to set RNG state via string stream is kind of cumbersome. Would it make sense to create a struct like `llama_rng` that just packages a random number generator, add it to the sampling context, and pass it for sampling?

In any case, what is the intended scope and functionality of saving/loading the state of `llama_context`? Should it include the sampling RNG state?

Reply:
> Would it make sense to create a struct like llama_rng that just packages a random number generator, add it to the sampling context, and pass it for sampling?

The `struct llama_sampling_context` is basically that - it should contain the full state related to sampling, so it makes sense to also move the RNG state there.

I don't think the RNG state should be stored with the `llama_context` state

Reply:
>The struct llama_sampling_context is basically that - it should contain the full state related to sampling, so it makes sense to also move the RNG state there.

What I meant specifically: on this PR I'm passing an instance of `std::mt19337` via `void *` but that's kind of a bad solution. Should I change this to a struct `llama_rng` defined in `llama.cpp` so that the sampling code can pass that instead of `void *`? Or should this PR be shelved until the sampling code is overhauled? To be clear, I am **not** interested in doing this, my primary goal is to get a baseline reproducible behavior that I can test my implementation in https://github.com/ggerganov/llama.cpp/pull/6828 against.

Reply:
You can move the `llama_sample_token_with_rng()` call to the internal C++ API over here:

https://github.com/ggerganov/llama.cpp/blob/4e96a812b3ce7322a29a3008db2ed73d9087b176/llama.h#L1073-L1076

Add a comment that this it is a temporary workaround. In the future, we'll update the sampling API to use `llama_sampling_context` as planned in #5214